### PR TITLE
PolyLoss

### PR DESCRIPTION
**Pull request by stdereka**
* Add PolyLoss implementation
* Add PolyLoss template config
* Bump up Ubuntu version in Dockerfile
* Fix protobuf version

### CODE REVIEW

File: mdn-metric/configs/exact/templates/template-polyloss.yaml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+criterion_params:
+  xent_weight: 0.0
+  polyloss_weight: 1.0
+  _hopt:
+    polyloss_epsilon:
+      min: 0
+      max: 2
```
Comment (ivan-chai) [inline mdn-metric/configs/exact/templates/template-polyloss.yaml]: max: 5?

File: mdn-metric/src/mdn_metric/criterion/polyloss.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        pt = torch.sum(labels_onehot * F.softmax(logits, dim=-1), dim=-1)
+        ce = F.cross_entropy(input=logits,
+                             target=labels,
+                             reduction='none',
+                             weight=self.weight)
+        poly1 = ce + self.epsilon * (1 - pt)
+        if self.reduction == "mean":
+            poly1 = poly1.mean()
+        elif self.reduction == "sum":
+            poly1 = poly1.sum()
```
Comment (ivan-chai) [inline mdn-metric/src/mdn_metric/criterion/polyloss.py:43]: else: assert == "none"

File: mdn-metric/src/mdn_metric/criterion/polyloss.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        """
+        Forward pass
+        :param logits: tensor of shape [N, num_classes]
+        :param labels: tensor of shape [N]
+        :return: poly cross-entropy loss
+        """
+        num_classes = logits.size(-1)
+        labels_onehot = F.one_hot(labels, num_classes=num_classes).to(device=logits.device,
+                                                                      dtype=logits.dtype)
+        pt = torch.sum(labels_onehot * F.softmax(logits, dim=-1), dim=-1)
```
Comment (ivan-chai) [inline mdn-metric/src/mdn_metric/criterion/polyloss.py:34]: softmax.take_along_dim(labels.unsqueeze(-1), dim=-1).squeeze(-1)

Comment (ivan-chai) [reply mdn-metric/src/mdn_metric/criterion/polyloss.py:34]: Не придется с one_hot мутить.

**Review by ivan-chai — CHANGES_REQUESTED** [review]

File: mdn-metric/configs/exact/templates/template-polyloss.yaml (hunk 1 above)
Comment (stdereka) [reply mdn-metric/configs/exact/templates/template-polyloss.yaml]: Done

File: mdn-metric/src/mdn_metric/criterion/polyloss.py (hunk 2 above)
Comment (stdereka) [reply mdn-metric/src/mdn_metric/criterion/polyloss.py:43]: Done

File: mdn-metric/src/mdn_metric/criterion/polyloss.py (hunk 3 above)
Comment (stdereka) [reply mdn-metric/src/mdn_metric/criterion/polyloss.py:34]: А может ну его?) Так более читаемо, по производительности нормально
